packages/oasgraph/src/resolver_builder.ts

```diff
@@ -181,6 +185,44 @@ export function getResolver ({
       qs: query
     }
 
+    /**
```

Comment:
I am concerned about `deepmerge`. I wonder if it is necessary. `deepmerge` is best suited for complex objects and because options, especially at this point in the code, only contains strings or hashes of strings, I think we can do the same amount of work by just using `Object.assign()`. 

As a side note, I am also not sure if I agree that `requestOptions` should have priority. 

Something along the lines of the following (I wrote it quickly so forgive me if there are mistakes)

```
let options: RequestOptions 
if (requestOptions) {
  options = requestOptions

  options['method'] = operation.method
  options['url'] = url
  if (options.headers) {
    Object.assign(options.headers, headers)
  } else {
    options['headers'] = headers
  }
  if (options.qs) {
    Object.assign(options.qs, query)
  } else {
    options['qs'] = query
  }
} else {
  options = {
    method: operation.method,
    url: url,
    headers: headers,
    qs: query
  }
}
```

We would like to minimize our dependencies so if we can avoid using them without writing too much code, then I think we should. Let me know your thoughts.

Reply:
Ha ! In my original work I used Object.assign(options, requestOptions) ! But doing so you would overwrite headers so I thought that for headers a "combine arrays" approach will be better than overwriting. Hence the deepmerge.

But with your approach you avoid this by explicitly assigning the headers, so I'm fine with it :)

Regarding your snippet, I'm fine with the approach ie. giving priority to options already set.

So if there is no other remarks I will amend the PR with a code similar to your snippet.

Reply:
I would like @ErikWittern to weight in but other than these changes, I am happy with what we have.

Reply:
Ok, let's wait some more discussion then :)

Reply:
I like the agreed-upon approach, where priority of fields is considered per case. I think line 3 should be `options = {...requestOptions}` (or use `Object.assign`) to not permanently overwrite `requestOptions` in the following lines?!

Reply:
Hmmm I think the same fields will be overwritten in all cases so it doesn't matter if we preserve `requestOptions`. I suppose it doesn't hurt just to do it...

```
let options: RequestOptions
if (requestOptions) {
  options = {...requestOptions}

  options['method'] = operation.method
  options['url'] = url
  if (options.headers) {
    Object.assign(options.headers, headers)
  } else {
    options['headers'] = headers
  }
  if (options.qs) {
    Object.assign(options.qs, query)
  } else {
    options['qs'] = query
  }
} else {
  options = {
    method: operation.method,
    url: url,
    headers: headers,
    qs: query
  }
}

```

Reply:
@thejibz If you can do this change and add the documentation like I mentioned in the PR comment, I think we'll be ready to accept!

Reply:
@Alan-Cha , on my way for the PR :) Hope to deliver today.

Reply:
> I like the agreed-upon approach, where priority of fields is considered per case. I think line 3 should be `options = {...requestOptions}` (or use `Object.assign`) to not permanently overwrite `requestOptions` in the following lines?!

Sounds good, doesn't work. Sorry for this "trait d'humour", in fact the Node Request's Options isn't fully compatible with the OASGraph "RequestOptions". E.g, in Node Request: "method" is optional, url is of type Url and not String...

So I think about amend the PR in the way that "RequestOptions" is no longer defined by OASGraph but taken directly from Node Request's Options type. Let me know if it works out for you ! The other way around is modifying "RequestOptions" to fully respect Node Request's options

Reply:
The could would be:
```
import * as NodeRequest from 'request'
// ...
let options: NodeRequest.OptionsWithUrl
    if (requestOptions) {
      options = { ...requestOptions }

      options['method'] = operation.method
      options['url'] = url
      if (options.headers) {
        Object.assign(options.headers, headers)
      } else {
        options['headers'] = headers
      }
      if (options.qs) {
        Object.assign(options.qs, query)
      } else {
        options['qs'] = query
      }
    } else {
      options = {
        method: operation.method,
        url: url,
        headers: headers,
        qs: query
      }
    }
```

Reply:
This is a great change! Thank you for doing the investigation work!